### ISSUE TITLE
ENH: High-dimensional LinearNDInterpolator

### ISSUE THREAD

**Issue by thchang** [scipy/scipy#19547]
### Is your feature request related to a problem? Please describe.

Scipy's ``LinearNDInterpolator`` function computes a Delaunay triangulation using the qhull algorithm, then uses it to perform Barycentric interpolation within each cell.

Qhull is a fast and robust algorithm, but the number of elements in the Delaunay triangulation (any mesh really) grows exponentially large with the dimension, making this solution unrealistic for d>4 dimension, even with just moderately sized data sets.

### Describe the solution you'd like.

An alternative approach, is to only compute the simplex containing each interpolation point.  This is generally faster whenever d>4, as long as the user only needs a polynomially-sized number of interpolation values (thus we don't end up needing to compute the entire mesh).

One of our papers proposes and implements such an approach.  It's based on a linear programming formulation of the problem, but with additional geometric checks and features to make sure that it is extremely robust (e.g., backward stable in a sense).

We have a Fortran driver published in ACM TOMS under MIT License:  https://dl.acm.org/doi/abs/10.1145/3422818

We have shown robustness on a wide variety of applications and scaled up to hundred-dimensional interpolation problems.

If there is interest, I would love to help get this integrated into scipy.interpolate.LinearNDInterpolator, perhaps we could switch off between this approach and the current approach using Qhull whenever the dimension exceeds some threshold, e.g,. 4d.

### Describe alternatives you've considered.

I believe the LP-based formulation has been known, but un-published for many years before us.  However, it is not so easy to implement this approach in a way that is robust to geometric degeneracy (i.e., the approach flat-out fails for many real-world data sets).

Our method requires the basic solution to the LP.  IP methods can't give the basic solution, so only simplex methods could be used.

Some open-source dense simplex-based LP solvers could be slightly faster on randomly generated data sets, but all that I have tested will fail for ill-conditioned and degenerate data sets (e.g., grid aligned data), since this results in degeneracy of the LP formulation.

### Additional context (e.g. screenshots, GIFs)

For reference, the following figures from the linked paper show scaling of our approach and various alternatives from the literature, including Qhull.

Time requirements of other approaches from the literature for computing the entire Delaunay triangulation in medium dimensions:

<img width="1542" alt="image" src="https://github.com/scipy/scipy/assets/17115193/68aac536-48dc-4aa2-96fb-ba1f007122fb">

Time requirements of our approach in much higher dimensions, for computing a single interpolation point (not the entire mesh):

<img width="1585" alt="image" src="https://github.com/scipy/scipy/assets/17115193/bf009e5f-eb67-4598-bfe0-a06b5962e153">

**Comment by lucascolley** [comment]
Hi @thchang , thanks for the proposal! I'm not an `interpolate` resident but this sounds interesting - could you send a proposal on the [mailing list](https://mail.python.org/archives/list/scipy-dev@python.org/) (with the key points here and linking to this issue) as stated in [the 'Contributing new code' section of our developer documentation](https://scipy.github.io/devdocs/dev/hacking.html#contributing-new-code)?

**Comment by ev-br** [comment]
This definitely sounds interesting! The speedups are impressive indeed.
ISTM what you're doing here is roughly a sort of sparse / lazy triangulation. Definitely very non-trivial to implement in a robust way! 

That said, here's some fine print:
- As Lucas said, it's best to discuss proposed enhancement on the scipy-dev mailing list. One thing you'd need to argue for inclusion into scipy is that the feature is of broad interest. The paper you linked has 4 citations from 2020, so there probably is some other metric / indicator / reasoning? We are fully aware that citations by themselves are not _the_ way to measure... almost anything; so I'd be interested to hear your arguments on the mailing list. 
- Off the cuff, it seems to be quite niche --- and quite a life-saver when Qhull runs out of steam of course. The curse of dimensionality is very real, the main question is whether this is better suited for a general-purpose library like SciPy or a more specialized package.
- Implementation details: there is going to be quite a push-back for including more Fortran code into SciPy. Again, this is not to criticize your choice of the implementation language, at all. It's just that Fortran/Python glue brings quite a headache for our build and distribution. You'd need to argue about that, too. I suspect rewriting the core algorithm in C/C++/Cython is not something you'd fancy doing.

TL;DR: The proposed enhancement looks very impressive and possibly in scope. What's not clear at the moment is whether to include it to scipy or release as a standalone package (ideally, a drop-in replacement for LinearNDInterpolator?). We can probably advise on how to implement python bindings if you go the separate package route.

**Comment by thchang** [comment]
Thank-you all for your feedback,

I understand the concerns.  I will discuss with my collaborators and reach to to the mailing list if needed.

We actually do have some python bindings, but they are not currently in the style of scipy.interpolate, also I don't have experience with wheels, so it requires a fortran compiler be pre-installed which can be called to build on import.

From our perspective, it would be nice to improve accessibility of what we've done, possibly through inclusion in scipy, but also possibly by improving our python bindings to match the scipy style.  I certainly understand the concerns on your end though.

As I said, will think on your feedback and send a proposal to the dev mailing list if needed.

**Comment by j-bowhay** [comment]
Feel free to leave the issue open in the meantime to continue discussions on and for visibility. Having hit the limitations of Qhull in my personal work I find this proposal interesting.
